frmts/gtiff/tif_jxl.c

```diff
@@ -599,23 +721,105 @@ JXLPostEncode(TIFF* tif)
+
+#ifdef HAVE_JxlExtraChannels
+        if (td->td_planarconfig == PLANARCONFIG_CONTIG &&
+            (basic_info.num_extra_channels > 1 ||
+             (basic_info.num_extra_channels == 1 && !bAlphaEmbedded)))
+        {
+            main_size = (sp->uncompressed_size / td->td_samplesperpixel);
+            int nMainChannels = basic_info.num_color_channels;
+            if(bAlphaEmbedded) nMainChannels++;
+            main_size *= nMainChannels;
+            main_buffer = _TIFFmalloc(main_size);
+            int outChunkSize = td->td_bitspersample / 8 * nMainChannels;
+            int inStep = nBytesPerSample * td->td_samplesperpixel;
+            uint8_t *cur_outbuffer=main_buffer;
+            uint8_t *cur_inbuffer=sp->uncompressed_buffer;
+            for( ; cur_outbuffer-main_buffer<main_size ; cur_outbuffer+=outChunkSize, cur_inbuffer+=inStep  ) {
+                memcpy(cur_outbuffer,cur_inbuffer,outChunkSize);
+            }
+            JxlExtraChannelInfo extra_channel_info;
+            JxlEncoderInitExtraChannelInfo(JXL_CHANNEL_OPTIONAL,
+                                           &extra_channel_info);
+            extra_channel_info.bits_per_sample = basic_info.bits_per_sample;
+            extra_channel_info.exponent_bits_per_sample = basic_info.exponent_bits_per_sample;
+            for(int iChannel=nMainChannels; iChannel<td->td_samplesperpixel; iChannel++)
+            {
+                if (JXL_ENC_SUCCESS != JxlEncoderSetExtraChannelInfo(
+                    enc,
+                    (bAlphaEmbedded)?iChannel-nMainChannels+1:iChannel-nMainChannels, &extra_channel_info))
+                {
+                    TIFFErrorExt(tif->tif_clientdata, module,
+                                 "JxlEncoderSetExtraChannelInfo(%d) failed",iChannel);
+                    JxlEncoderDestroy(enc);
+                    return 0;
+                }
+            }
+        }
+#endif
+        
 
-        if( JxlEncoderAddImageFrame(opts, &format, sp->uncompressed_buffer,
-                                    sp->uncompressed_size) != JXL_ENC_SUCCESS )
+        int retCode = JxlEncoderAddImageFrame(opts, &format, main_buffer, main_size);
+        //cleanup now
+        if(main_buffer!=sp->uncompressed_buffer) {
+            _TIFFfree(main_buffer);
+        }
+        if( retCode != JXL_ENC_SUCCESS )
         {
             TIFFErrorExt(tif->tif_clientdata, module,
                          "JxlEncoderAddImageFrame() failed");
             JxlEncoderDestroy(enc);
             return 0;
         }
+
+#ifdef HAVE_JxlExtraChannels
+        if (td->td_planarconfig == PLANARCONFIG_CONTIG &&
+            (basic_info.num_extra_channels > 1 ||
+             (basic_info.num_extra_channels == 1 && !bAlphaEmbedded)))
+        {
+            int nMainChannels = basic_info.num_color_channels;
+            if(bAlphaEmbedded) nMainChannels++;
+            int extra_channel_size = (sp->uncompressed_size / td->td_samplesperpixel);
```

Comment:
this is actually the total number of samples per pixel. should probably be
```suggestion
            int extra_channel_size = td->td_samplesperpixel - nMainChannels;
```

Reply:
extra_channel_size is the number of bytes needed to hold a single channel

Reply:
makes sense. ignore my comment